### PR TITLE
GRIDEDIT-1244 Split row/column for unstructured meshes

### PR DESCRIPTION
**Pull request by BillSenior**

### CODE REVIEW

File: libs/MeshKernel/include/MeshKernel/SplitRowColumnOfMesh.hpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                           std::vector<UInt>& edgesToDelete) const;
+
+        /// @brief Collect the ID's of the edges and elements to be split.
+        void CollectElementsToSplit(const Mesh2D& mesh, const UInt edgeId, std::vector<UInt>& elementIds, std::vector<UInt>& edgeIds) const;
+
+        /// @brief Determine if it may be possible to split the edge
+        ///
+        /// If two elements are attached to edge then either must be a quadrilateral,
+        /// otherwise the single attached element must be quadrilateral.
+        bool MayBeSplit(const Mesh2D& mesh, const UInt edgeId) const;
```
Comment (lucacarniato) [inline libs/MeshKernel/include/MeshKernel/SplitRowColumnOfMesh.hpp:96]: I would rename it CanBeSplit?

File: libs/MeshKernel/src/Mesh.cpp [hunk 2]
```diff
@@ -1032,6 +1032,72 @@ meshkernel::UInt Mesh::GetNumValidEdges() const
...
+    }
+
+    if (edgeId >= GetNumEdges())
+    {
+        throw ConstraintError("edge id is greater than the number of edges: {} >= {}", edgeId, GetNumEdges());
+    }
+
+    const std::vector<UInt>& edgeIds = m_facesEdges[elementId];
+
+    // TODO use Operations::FindIndex when curvilinear grid from splines has been added to master
```
Comment (lucacarniato) [inline libs/MeshKernel/src/Mesh.cpp:1054]: remove todo?

File: libs/MeshKernel/src/SplitRowColumnOfMesh.cpp [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+void meshkernel::SplitRowColumnOfMesh::SplitElement(Mesh2D& mesh,
+                                                    const UInt elementId,
+                                                    const UInt edgeId,
+                                                    UInt& newNode,
+                                                    CompoundUndoAction& undoActions,
+                                                    std::vector<UInt>& edgesToDelete) const
+{
+    const Edge edgeNode = mesh.GetEdge(edgeId);
+    const std::array<UInt, 2>& edgeFace = mesh.m_edgesFaces[edgeId];
+    UInt previousElementId = edgeFace[0] + edgeFace[1] - elementId;
```
Comment (lucacarniato) [inline libs/MeshKernel/src/SplitRowColumnOfMesh.cpp:174]: use const

File: libs/MeshKernel/src/SplitRowColumnOfMesh.cpp [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+
+    partialElementIds.reserve(initialNumberOfElements);
+    partialEdgeIds.reserve(initialNumberOfElements);
+
+    // Loop over each side of the edge
+    for (UInt i = 0; i < 2; ++i)
+    {
+        if (!IsValid(mesh.m_edgesFaces[edgeId][i]))
+        {
+            // There is no elememnt on the other side of the edge.
```
Comment (lucacarniato) [inline libs/MeshKernel/src/SplitRowColumnOfMesh.cpp:267]: elememnt => element

File: libs/MeshKernel/src/SplitRowColumnOfMesh.cpp [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+
+bool meshkernel::SplitRowColumnOfMesh::IsQuadrilateral(const Mesh2D& mesh, const UInt elementId) const
+{
+    return IsValid(elementId) && mesh.m_numFacesNodes[elementId] == constants::geometric::numNodesInQuadrilateral;
+}
+
+void meshkernel::SplitRowColumnOfMesh::CollectElementsToSplit(const Mesh2D& mesh, const UInt edgeId, std::vector<UInt>& elementIds, std::vector<UInt>& edgeIds) const
+{
+    const UInt firstElementId = mesh.m_edgesFaces[edgeId][0] != constants::missing::uintValue ? mesh.m_edgesFaces[edgeId][0] : mesh.m_edgesFaces[edgeId][1];
+    const UInt initialNumberOfElements = static_cast<UInt>(std::sqrt(static_cast<double>(mesh.GetNumNodes())));
```
Comment (lucacarniato) [inline libs/MeshKernel/src/SplitRowColumnOfMesh.cpp:254]: GetNumFaces?

Comment (lucacarniato) [inline libs/MeshKernel/src/SplitRowColumnOfMesh.cpp:254]: initialNumberOfFaces?

File: libs/MeshKernel/src/SplitRowColumnOfMesh.cpp [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+
+    std::vector<UInt> partialElementIds;
+    std::vector<UInt> partialEdgeIds;
+
+    partialElementIds.reserve(initialNumberOfElements);
+    partialEdgeIds.reserve(initialNumberOfElements);
+
+    // Loop over each side of the edge
+    for (UInt i = 0; i < 2; ++i)
```
Comment (lucacarniato) [inline libs/MeshKernel/src/SplitRowColumnOfMesh.cpp:263]: create a function and call it twice? this could reduce the cyclomatic complexity a bit

**Review by lucacarniato — CHANGES_REQUESTED** [review]

File: libs/MeshKernelApi/include/MeshKernelApi/MeshKernel.hpp [hunk 7]
```diff
@@ -1450,6 +1450,15 @@ namespace meshkernelapi
         /// @returns Error code
         MKERNEL_API int mkernel_mesh2d_set(int meshKernelId, const Mesh2D& mesh2d);
 
+        /// @brief Sets the meshkernel::Mesh2D state
+        /// @param[in] meshKernelId The id of the mesh state
```
Comment (BillSenior) [inline libs/MeshKernelApi/include/MeshKernelApi/MeshKernel.hpp:1454]: Correct comment

**Review by lucacarniato — APPROVED** [review]

**Pull request merged**